app/server/ruby/bin/daemon.rb

```diff
@@ -179,63 +203,53 @@ def initialize
         Util.log "Spider Server process has completed"
       end
 
+      def udp_zombie_kill_switch(port_num, exit_token)
+        kill_switch = Promise.new
+        queue = Queue.new
 
-      # This is the Zombie Kill Switch
-      #
-      # It is a separate thread which calls exit (therefore triggering
-      # the at_exit block which in turn calls
-      # cleanup_any_running_processes).
-      #
-      # There are the following exit conditions:
-      #
-      # Initial Connection Timeout - if nothing connects within 5 seconds,
-      #                              switch is activated
-      #
-      # Connection Lost            - if the connection is broken or there's
-      #                              a timeout - waiting for the next
-      #                              keep-alive message.
-      #
-      def spawn_zombie_kill_switch(port_num, &blk)
-        # Define how long to wait for an initial connection and between keep-alive messages.
-        connect_timeout = 5
-        recv_timeout = 30
-
-        # Spawn TCP server
-        keep_alive_server = TCPServer.new "127.0.0.1", port_num
-
-        Thread.new do
-          begin
-            Util.log "Waiting for GUI on port #{port_num}...."
+        kill_switch_thread = Thread.new do
+          attempts = 0
+          max_attempts = 4
 
-            unless IO.select([keep_alive_server], nil, nil, connect_timeout)
-              Util.log "Critical Error. Unable to connect to GUI process on TCP port #{port_num}"
-              @safe_exit.exit
+          loop do
+            Kernel.sleep 5
+            if queue.empty?
+              attempts += 1
+            else
+              attempts = 0
+              queue.clear
             end
+            kill_switch.deliver!(true) if attempts > max_attempts
+          end
+        end
 
-            client = keep_alive_server.accept
-            Util.log "Connected to GUI!"
+        server = SonicPi::OSC::UDPServer.new(port_num) do |address, args, sender_addrinfo|
+          Util.log "Kill switch ##{port_num} Received UDP data #{[address, args, sender_addrinfo].inspect}"
+        end
+
+        server.add_method("/daemon/keep-alive") do |args|
+          queue << true
+        end
 
-            while IO.select([client], nil, nil, recv_timeout) && received_data = client.gets
-              # Don't do anything with incoming data.
-              # Its presence signifies things are still OK.
-              # For debug:
-              # Util.log "RCV #{received_data}"
+        if exit_token
+          server.add_method("/daemon/exit") do |args|
+            if args[0] && args[0] == exita_token
```

Comment:
Was this a typo and supposed to be just `exit_token`?

```suggestion
            if args[0] && args[0] == exit_token
```